File: app/src/main/java/com/example/bridgeapp/logic/PointStageState.java
package com.example.bridgeapp.logic;

import android.os.Parcel;
import android.os.Parcelable;

import com.example.bridgeapp.structure.ContractObject;
import com.example.bridgeapp.structure.HandObject;
import com.example.bridgeapp.structure.PointObject;
import com.example.bridgeapp.util.CardSuit;

import java.util.ArrayList;
import java.util.List;

/*
NOTE:   The only score you cannot get from this class is a bonus for winning rubber (2 games).
*/

public class PointStageState implements Parcelable {

    private final List<PointObject> pointsBelowLine;
    private final List<PointObject> pointsAboveLine;

    private final List<PointObject> pointsDefence;

    public PointStageState() {
        pointsBelowLine = new ArrayList<>();
        pointsAboveLine = new ArrayList<>();

        pointsDefence = new ArrayList<>();
    }

    protected PointStageState(Parcel in) {
        pointsBelowLine = in.createTypedArrayList(PointObject.CREATOR);
        pointsAboveLine = in.createTypedArrayList(PointObject.CREATOR);
        pointsDefence = in.createTypedArrayList(PointObject.CREATOR);
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeTypedList(pointsBelowLine);
        dest.writeTypedList(pointsAboveLine);
        dest.writeTypedList(pointsDefence);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    public static final Creator<PointStageState> CREATOR = new Creator<PointStageState>() {
        @Override
        public PointStageState createFromParcel(Parcel in) {
            return new PointStageState(in);
        }

        @Override
        public PointStageState[] newArray(int size) {
            return new PointStageState[size];
        }
    };

    public int scoreContract(int tricksTaken, ContractObject contract) {
        int points = 0;
        if (checkContract(contract, tricksTaken)) {
            if(contract.getSuit().ordinal() >=2){
                points += 30 * contract.getNumber();
                if(contract.getSuit().ordinal() >=4)
                    points += 10;
            }
            else {
                points += 20 * contract.getNumber();
            }
            if(contract.getDoubleVal() > 0)
                points *= contract.getDoubleVal() * 2;
        }
        return points;
    }

    public int scoreOvertricks(int tricksTaken, ContractObject contract, boolean vulnerable){
        int points = 0;
        if (checkContract(contract, tricksTaken)) {
            if(contract.getDoubleVal() == 0){
                if(contract.getSuit().ordinal() >=2){
                points += 30 * ((tricksTaken - 6) - contract.getNumber());
                if(contract.getSuit().ordinal() >=4)
                    points += 10;
                }
                else {
                    points += 20 * ((tricksTaken - 6) - contract.getNumber());
                }
            }
            else{
                points = 100 * contract.getDoubleVal() * ((tricksTaken - 6) - contract.getNumber());
                if(vulnerable){
                    points *= 2;
                }
            }
        }
        return points;
    }

    public int scoreMishaps(int tricksTaken, ContractObject contract, boolean vulnerable){
        int points = 0;
        if (!checkContract(contract, tricksTaken)) {
            if(contract.getDoubleVal() > 0){
                points += 300;
                points *= (contract.getNumber() + 6) - tricksTaken;
                if(!vulnerable){
                    points -= 300;
                }
                points -= 100;
                points *= contract.getDoubleVal();
            }
            else{
                points = 50 * ((contract.getNumber() + 6) - tricksTaken);
                if(vulnerable){
                    points *= 2;
                }
            }
        }
        return points;
    }


    public int getHonorBonus(HandObject hand, CardSuit trump){
        if (hand.honorBonus(trump)==1) return 100;
        else if (hand.honorBonus(trump)==2) return 150;
        else return 0;
    }

    public int getDoubleBonus(int doubleValue){
        return 50 * doubleValue;
    }

    public int getSlamBonus(int tricksTaken, ContractObject contract, boolean vulnerable){
        if(checkContract(contract, tricksTaken)){
            int points = 0;
            if(contract.getNumber()+6 >=12){
                points += 500;
                if(vulnerable)
                    points += 250;
                if(contract.getNumber()+6 == 13)
                    points *= 2;
            }
            return points;
        }
        return 0;
    }

    public int getRubberBonus(boolean didWin, boolean enemyVulnerable){
        if(didWin){
            if(enemyVulnerable)
                return 500;
            else
                return 700;
        }
        return 0;
    }

    public static boolean areTricksCorrect(int tricks){
        return (tricks >= 0 && tricks <= 13);
    }

    public static boolean checkContract(ContractObject contract, int tricksTaken){
        return tricksTaken - 6 >= contract.getNumber();
    }

    private void setPointsBelowLine(int tricksTaken, ContractObject contract){
        pointsBelowLine.add(new PointObject(scoreContract(tricksTaken, contract), "Contract"));
    }

    private void setPointsAboveLine(int tricksTaken, ContractObject contract, boolean vulnerable, boolean enemyVulnerable, boolean isWinner, HandObject[] hands){
        if(scoreOvertricks(tricksTaken, contract, vulnerable) > 0)
            pointsAboveLine.add(new PointObject(scoreOvertricks(tricksTaken, contract, vulnerable), "Overtricks"));
        if(checkContract(contract, tricksTaken) && getDoubleBonus(contract.getDoubleVal()) > 0)
            pointsAboveLine.add(new PointObject(getDoubleBonus(contract.getDoubleVal()),(contract.getDoubleVal()>1 ? "Redouble": "Double")));
        if(checkContract(contract, tricksTaken) && getSlamBonus(tricksTaken, contract, vulnerable) > 0)
            pointsAboveLine.add(new PointObject(getSlamBonus(tricksTaken, contract, vulnerable),(tricksTaken > 12 ? "Grand slam":"Small slam")));
        if(getHonorBonus(hands[contract.getPlayer()], contract.getSuit()) + getHonorBonus(hands[(contract.getPlayer()+2)%4], contract.getSuit()) > 0)
            pointsAboveLine.add(new PointObject(getHonorBonus(hands[contract.getPlayer()], contract.getSuit()) + getHonorBonus(hands[(contract.getPlayer()+2)%4], contract.getSuit()), "Honors"));
        if(getRubberBonus(isWinner, enemyVulnerable) > 0)
              pointsAboveLine.add(new PointObject(getRubberBonus(isWinner, enemyVulnerable), "Rubber won"));
    }

    private void setDefencePoints(int tricksTaken, ContractObject contract, boolean vulnerable, boolean enemyVulnerable, boolean isWinner, HandObject[] hands){
        if(scoreMishaps(tricksTaken, contract, vulnerable) > 0)
            pointsDefence.add(new PointObject(scoreMishaps(tricksTaken, contract, vulnerable), "Mishaps"));
        if(getHonorBonus(hands[(contract.getPlayer()+1)%4], contract.getSuit()) > 0)
            pointsDefence.add(new PointObject(getHonorBonus(hands[(contract.getPlayer()+1)%4], contract.getSuit()) + getHonorBonus(hands[(contract.getPlayer()+3)%4], contract.getSuit()), "Honors"));
        if(getRubberBonus(isWinner, enemyVulnerable) > 0)
            pointsAboveLine.add(new PointObject(getRubberBonus(isWinner, enemyVulnerable), "Rubber won"));
    }

    public void setPoints(int tricksTaken, ContractObject contract, boolean vulnerable, boolean enemyVulnerable, int winner, HandObject[] hands){
        pointsDefence.clear();
        pointsAboveLine.clear();
        pointsBelowLine.clear();

        setPointsBelowLine(tricksTaken, contract);
        setPointsAboveLine(tricksTaken, contract, vulnerable, enemyVulnerable, contract.getPlayer()%2 == winner, hands);
        setDefencePoints(tricksTaken, contract, vulnerable, vulnerable, (contract.getPlayer()+1)%2 == winner, hands);
    }

    public List<PointObject> getPointsBelowLine() {
        return pointsBelowLine;
    }

    public List<PointObject> getPointsAboveLine() {
        return pointsAboveLine;
    }

    public List<PointObject> getPointsDefence() {
        return pointsDefence;
    }

}

File: app/src/main/java/com/example/bridgeapp/logic/PointStageState.kt
package com.example.bridgeapp.logic

import android.os.Parcel
import android.os.Parcelable
import com.example.bridgeapp.structure.ContractObject
import com.example.bridgeapp.structure.HandObject
import com.example.bridgeapp.structure.PointObject
import com.example.bridgeapp.util.CardSuit
import java.util.*

/*
NOTE:   The only score you cannot get from this class is a bonus for winning rubber (2 games).
*/
class PointStageState : Parcelable {
    val pointsBelowLine: MutableList<PointObject>?
    val pointsAboveLine: MutableList<PointObject>?
    val pointsDefence: MutableList<PointObject>?

    constructor() {
        pointsBelowLine = ArrayList()
        pointsAboveLine = ArrayList()
        pointsDefence = ArrayList()
    }

    private constructor(`in`: Parcel) {
        pointsBelowLine = `in`.createTypedArrayList<PointObject>(PointObject.CREATOR)!!.toMutableList()
        pointsAboveLine = `in`.createTypedArrayList<PointObject>(PointObject.CREATOR)!!.toMutableList()
        pointsDefence = `in`.createTypedArrayList<PointObject>(PointObject.CREATOR)!!.toMutableList()
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeTypedList(pointsBelowLine)
        dest.writeTypedList(pointsAboveLine)
        dest.writeTypedList(pointsDefence)
    }

    override fun describeContents(): Int {
        return 0
    }

    fun scoreContract(tricksTaken: Int, contract: ContractObject): Int {
        var points = 0
        if (checkContract(contract, tricksTaken)) {
            if (contract.suit!!.ordinal >= 2) {
                points += 30 * contract.number
                if (contract.suit.ordinal >= 4) points += 10
            } else {
                points += 20 * contract.number
            }
            if (contract.doubleVal > 0) points *= contract.doubleVal * 2
        }
        return points
    }

    fun scoreOvertricks(tricksTaken: Int, contract: ContractObject, vulnerable: Boolean): Int {
        var points = 0
        if (checkContract(contract, tricksTaken)) {
            if (contract.doubleVal == 0) {
                if (contract.suit!!.ordinal >= 2) {
                    points += 30 * (tricksTaken - 6 - contract.number)
                    if (contract.suit.ordinal >= 4) points += 10
                } else {
                    points += 20 * (tricksTaken - 6 - contract.number)
                }
            } else {
                points = 100 * contract.doubleVal * (tricksTaken - 6 - contract.number)
                if (vulnerable) {
                    points *= 2
                }
            }
        }
        return points
    }

    fun scoreMishaps(tricksTaken: Int, contract: ContractObject, vulnerable: Boolean): Int {
        var points = 0
        if (!checkContract(contract, tricksTaken)) {
            if (contract.doubleVal > 0) {
                points += 300
                points *= contract.number + 6 - tricksTaken
                if (!vulnerable) {
                    points -= 300
                }
                points -= 100
                points *= contract.doubleVal
            } else {
                points = 50 * (contract.number + 6 - tricksTaken)
                if (vulnerable) {
                    points *= 2
                }
            }
        }
        return points
    }

    private fun getHonorBonus(hand: HandObject, trump: CardSuit?): Int {
        return if (hand.honorBonus(trump) == 1) 100 else if (hand.honorBonus(trump) == 2) 150 else 0
    }

    private fun getDoubleBonus(doubleValue: Int): Int {
        return 50 * doubleValue
    }

    private fun getSlamBonus(tricksTaken: Int, contract: ContractObject, vulnerable: Boolean): Int {
        if (checkContract(contract, tricksTaken)) {
            var points = 0
            if (contract.number + 6 >= 12) {
                points += 500
                if (vulnerable) points += 250
                if (contract.number + 6 == 13) points *= 2
            }
            return points
        }
        return 0
    }

    private fun getRubberBonus(didWin: Boolean, enemyVulnerable: Boolean): Int {
        return if (didWin) {
            if (enemyVulnerable) 500 else 700
        } else 0
    }

    private fun setPointsBelowLine(tricksTaken: Int, contract: ContractObject) {
        pointsBelowLine!!.add(PointObject(scoreContract(tricksTaken, contract), "Contract"))
    }

    private fun setPointsAboveLine(tricksTaken: Int, contract: ContractObject, vulnerable: Boolean, enemyVulnerable: Boolean, isWinner: Boolean, hands: Array<HandObject>) {
        if (scoreOvertricks(tricksTaken, contract, vulnerable) > 0) pointsAboveLine!!.add(PointObject(scoreOvertricks(tricksTaken, contract, vulnerable), "Overtricks"))
        if (checkContract(contract, tricksTaken) && getDoubleBonus(contract.doubleVal) > 0) pointsAboveLine!!.add(PointObject(getDoubleBonus(contract.doubleVal), if (contract.doubleVal > 1) "Redouble" else "Double"))
        if (checkContract(contract, tricksTaken) && getSlamBonus(tricksTaken, contract, vulnerable) > 0) pointsAboveLine!!.add(PointObject(getSlamBonus(tricksTaken, contract, vulnerable), if (tricksTaken > 12) "Grand slam" else "Small slam"))
        if (getHonorBonus(hands[contract.player], contract.suit) + getHonorBonus(hands[(contract.player + 2) % 4], contract.suit) > 0) pointsAboveLine!!.add(PointObject(getHonorBonus(hands[contract.player], contract.suit) + getHonorBonus(hands[(contract.player + 2) % 4], contract.suit), "Honors"))
        if (getRubberBonus(isWinner, enemyVulnerable) > 0) pointsAboveLine!!.add(PointObject(getRubberBonus(isWinner, enemyVulnerable), "Rubber won"))
    }

    private fun setDefencePoints(tricksTaken: Int, contract: ContractObject, vulnerable: Boolean, enemyVulnerable: Boolean, isWinner: Boolean, hands: Array<HandObject>) {
        if (scoreMishaps(tricksTaken, contract, vulnerable) > 0) pointsDefence!!.add(PointObject(scoreMishaps(tricksTaken, contract, vulnerable), "Mishaps"))
        if (getHonorBonus(hands[(contract.player + 1) % 4], contract.suit) > 0) pointsDefence!!.add(PointObject(getHonorBonus(hands[(contract.player + 1) % 4], contract.suit) + getHonorBonus(hands[(contract.player + 3) % 4], contract.suit), "Honors"))
        if (getRubberBonus(isWinner, enemyVulnerable) > 0) pointsAboveLine!!.add(PointObject(getRubberBonus(isWinner, enemyVulnerable), "Rubber won"))
    }

    fun setPoints(tricksTaken: Int, contract: ContractObject, vulnerable: Boolean, enemyVulnerable: Boolean, winner: Int, hands: Array<HandObject>) {
        pointsDefence!!.clear()
        pointsAboveLine!!.clear()
        pointsBelowLine!!.clear()
        setPointsBelowLine(tricksTaken, contract)
        setPointsAboveLine(tricksTaken, contract, vulnerable, enemyVulnerable, contract.player % 2 == winner, hands)
        setDefencePoints(tricksTaken, contract, vulnerable, vulnerable, (contract.player + 1) % 2 == winner, hands)
    }

    companion object {
        @JvmField val CREATOR: Parcelable.Creator<PointStageState?> = object : Parcelable.Creator<PointStageState?> {
            override fun createFromParcel(`in`: Parcel): PointStageState? {
                return PointStageState(`in`)
            }

            override fun newArray(size: Int): Array<PointStageState?> {
                return arrayOfNulls(size)
            }
        }

        fun areTricksCorrect(tricks: Int): Boolean {
            return tricks in 0..13
        }

        fun checkContract(contract: ContractObject, tricksTaken: Int): Boolean {
            return tricksTaken - 6 >= contract.number
        }
    }
}